### ISSUE TITLE
Remote COPC: if the server does not support range requests, QGIS freezes

### ISSUE THREAD

**Issue by wonder-sk** [qgis/QGIS#48304]
### What is the bug or the crash?

I tried to open a remote COPC file: https://github.com/PDAL/data/raw/master/autzen/autzen-classified.copc.laz

Apparently that server does not understand range requests, so when loading the layer, QGIS freezes completely and probably it's blocking all the way until the whole file would not get downloaded.

When reading COPC header, we should detect if server supports range requests (i.e. it responds with 206 http status) and if not, abort the request as soon as possible, and show an error to the user.

### Steps to reproduce the issue

1. open data sources manager, go to point clouds tab, select protocol and enter URL above
2. qgis freezes

### Versions

master

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by uclaros** [comment]
@wonder-sk @NEDJIMAbelgacem 
Can we rely on the reply header's `Accept-Ranges` attribute?
We could do a `HEAD` instead of a `GET` request then and not care about the length of a blocking request.

The specified example however shows another aspect of this issue: The provided link returns a 302 redirect to a server that *can* actually handle range requests!

`curl -I https://github.com/PDAL/data/raw/master/autzen/autzen-classified.copc.laz` returns:
```
HTTP/2 302 
server: GitHub.com
date: Fri, 03 Jun 2022 10:20:56 GMT
content-type: text/html; charset=utf-8
vary: X-PJAX, X-PJAX-Container, Turbo-Visit, Turbo-Frame, Accept-Encoding, Accept, X-Requested-With
permissions-policy: interest-cohort=()
access-control-allow-origin: https://render.githubusercontent.com
location: https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen-classified.copc.laz
cache-control: no-cache
strict-transport-security: max-age=31536000; includeSubdomains; preload
x-frame-options: deny
x-content-type-options: nosniff
x-xss-protection: 0
referrer-policy: no-referrer-when-downgrade
expect-ct: max-age=2592000, report-uri="https://api.github.com/_private/browser/errors"
content-security-policy: default-src 'none'; base-uri 'self'; block-all-mixed-content; child-src github.com/assets-cdn/worker/ gist.github.com/assets-cdn/worker/; connect-src 'self' uploads.github.com objects-origin.githubusercontent.com www.githubstatus.com collector.github.com raw.githubusercontent.com api.github.com github-cloud.s3.amazonaws.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com cdn.optimizely.com logx.optimizely.com/v1/events *.actions.githubusercontent.com wss://*.actions.githubusercontent.com online.visualstudio.com/api/v1/locations github-production-repository-image-32fea6.s3.amazonaws.com github-production-release-asset-2e65be.s3.amazonaws.com insights.github.com wss://alive.github.com; font-src github.githubassets.com; form-action 'self' github.com gist.github.com objects-origin.githubusercontent.com; frame-ancestors 'none'; frame-src render.githubusercontent.com viewscreen.githubusercontent.com notebooks.githubusercontent.com; img-src 'self' data: github.githubassets.com identicons.github.com github-cloud.s3.amazonaws.com secured-user-images.githubusercontent.com/ github-production-user-asset-6210df.s3.amazonaws.com *.githubusercontent.com; manifest-src 'self'; media-src github.com user-images.githubusercontent.com/; script-src github.githubassets.com; style-src 'unsafe-inline' github.githubassets.com; worker-src github.com/assets-cdn/worker/ gist.github.com/assets-cdn/worker/
content-length: 0
x-github-request-id: DCF4:0DCA:3083D0:3DBA35:6299E088
```

and `curl -I https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen-classified.copc.laz` actually returns:
```
HTTP/2 200 
cache-control: max-age=300
content-security-policy: default-src 'none'
content-type: application/octet-stream
etag: "db2d56cdfa058bffccdc5d6019dae2fc9c6a551df10a5523c06c76a3e25a27fa"
last-modified: Fri, 03 Jun 2022 10:24:36 GMT
strict-transport-security: max-age=31557600
timing-allow-origin: https://github.com
x-content-type-options: nosniff
x-frame-options: deny
x-xss-protection: 1; mode=block
x-github-request-id: DB30:98A1:28B4B1:33F5C4:6299E163
accept-ranges: bytes
date: Fri, 03 Jun 2022 10:24:38 GMT
via: 1.1 varnish
x-served-by: cache-sof1510051-SOF
x-cache: MISS
x-cache-hits: 0
x-timer: S1654251876.850637,VS0,VE2946
access-control-allow-origin: *
x-fastly-request-id: f69206709202ba54330baf5ff84cdb88bd53d055
expires: Fri, 03 Jun 2022 10:29:38 GMT
source-age: 3
vary: Authorization,Accept-Encoding
content-length: 81123042
```

Maybe we should handle redirects first?

**Comment by wonder-sk** [comment]
Good points @uclaros 

`Accept-Ranges` header should be hopefully reliable piece of information that server supports range queries, so in theory we could use that with a blocking HEAD request, to avoid more QEventLoops at the cost of doing one extra initial request...

As for redirects - it looks like QgsBlockingNetworkRequest already handles redirects automatically (including redirect loop detection) - so hopefully there's nothing to worry about...

**Comment by uclaros** [comment]
> As for redirects - it looks like QgsBlockingNetworkRequest already handles redirects automatically (including redirect loop detection) - so hopefully there's nothing to worry about...

Why does the provided url not work then? forwarder server does not support range, target does support

**Comment by NEDJIMAbelgacem** [comment]
> Good points @uclaros
> 
> `Accept-Ranges` header should be hopefully reliable piece of information that server supports range queries, so in theory we could use that with a blocking HEAD request, to avoid more QEventLoops at the cost of doing one extra initial request...
> 
> As for redirects - it looks like QgsBlockingNetworkRequest already handles redirects automatically (including redirect loop detection) - so hopefully there's nothing to worry about...

Yes, but this way the actual checking for range query support will be done in a seperate request before getting the header. From our recent call I argued that the checking for range query support should be done in a seperate function/request but sadly I didn't explain that well and ended up implementing your suggestion about doing this in the header fetching.

What I'm going to do now is revert the changes to the LazInfo class and add a new function that checks whether the server supports range queries (that function will do a blocking network HEAD request) and before fetching the header we will use that function to check for range query support.

Hopefully this makes sense and thanks to @uclaros for suggesting a HEAD request instead of GET.

**Comment by wonder-sk** [comment]
Fixed in https://github.com/qgis/QGIS/pull/48824

**Comment by tomkom** [comment]
Can I ask in which QGIS release this will be included? I am using 3.26.1 on Mac OSX, and tested with the typical test case of autzen-classified.copc.laz (to have a shareable doc):
https://objects.storage.unimelb.edu.au/4320_budjbimdata/COGs/autzen-classified.copc.laz

This is sitting on our institutional S3 compatible storage. Works fine with COGs.
I get: 
Unsupported Data Source: /vsicurl/https://objects.storage.unimelb.edu.au/4320_budjbimdata/COGs/autzen-classified.copc.laz is not a supported raster data source HTTP response code on https://objects.storage.unimelb.edu.au/4320_budjbimdata/COGs/autzen-classified.copc.laz.xml: 403

**Comment by tomkom** [comment]
Downloading the file and running pdal info on it, we get a normal expected output (I am not sure if pdal info can do /vsicurl/https.....: 
{
    “file_size”: 81123042,
    “filename”: “COPC/autzen-classified.copc.laz”,
    “now”: “2022-08-26T01:20:24+0000”,
    “pdal_version”: “2.4.2 (git-version: 3d466d)“,
    “reader”: “readers.copc”,
    “stats”:
    {
      “bbox”:
      {
        “EPSG:4326”:
...

**Comment by zexhuang** [comment]
Dear @NEDJIMAbelgacem 

bool QgsLazInfo::supportsRangeQueries( QUrl &url )
{
  QNetworkRequest nr( url );
  nr.setAttribute( QNetworkRequest::CacheLoadControlAttribute, QNetworkRequest::AlwaysNetwork );
  nr.setAttribute( QNetworkRequest::CacheSaveControlAttribute, false );
  nr.setRawHeader( "Range", "bytes=0-0" );
  QgsBlockingNetworkRequest req;
  QgsBlockingNetworkRequest::ErrorCode errCode = req.head( nr );
  QgsNetworkReplyContent reply = req.reply();

  QString acceptRangesHeader = reply.rawHeader( QStringLiteral( "Accept-Ranges" ).toLocal8Bit() );
  return errCode == QgsBlockingNetworkRequest::NoError && acceptRangesHeader.compare( QStringLiteral( "bytes" ), Qt::CaseSensitivity::CaseInsensitive ) == 0;
}

The above function checks whether the client url supports range queries. But this function return False values for the url: https://objects.storage.unimelb.edu.au/4320_budjbimdata/COPC/autzen-classified.copc.laz. 

curl -I https://objects.storage.unimelb.edu.au/4320_budjbimdata/COPC/autzen-classified.copc.laz returns:

HTTP/2 200 
content-length: 81123042
accept-ranges: bytes
last-modified: Fri, 26 Aug 2022 07:23:31 GMT
x-rgw-object-type: Normal
etag: "dbb36ebb301306feb94c5e313524492c-10"
x-amz-request-id: tx000000000000000983f48-0063087511-178a3e9c-unimelb-qh2
content-type: binary/octet-stream
date: Fri, 26 Aug 2022 07:24:01 GMT
set-cookie: SERVERID=rgw2-ceph3-qh2-new.lb.storage.unimelb.edu.au; path=/
cache-control: private

Could you please check this?

Many thanks!

**Comment by wonder-sk** [comment]
OK I can see the issue now. Your server returns `accept-ranges` header in lowercase and our code checks for `Accept-Ranges` header. Normally in `QNetworkReply` headers are treated case insensitive (https://code.woboq.org/qt5/qtbase/src/network/access/qnetworkrequest.cpp.html#1146), but the wrapper class `QgsNetworkReplyContent` returned from `QgsBlockingNetworkRequest` uses case sensitive header lookup and it fails.

@NEDJIMAbelgacem could you please fix it in QgsNetworkReplyContent code?

@tomkom for the time being, if you can force this http header on your server to be `Accept-Ranges`, then things should work...

**Comment by uclaros** [comment]
@wonder-sk I've already fixed that in `QgsLazInfo` in #49966

In `QgsNetworkReplyContent` the related methods are `rawHeader` - `rawHeaderList` - `hasRawHeader` so a case insensitive search would not really comply with the `Raw` part.

**Comment by wonder-sk** [comment]
Oh I see - github somehow failed for me to show the reference to https://github.com/qgis/QGIS/pull/49966 when I was looking at that.

It think the rawHeader() and hasRawHeader() should be case insensitive in QgsNetworkReplyContent - they are in QNetworkReply (see the link above to QNetworkReply implementation) and it is confusing to have different behavior in the QGIS class. The `Raw` part of `rawHeader()` really is there as a differentiation to `header()` which uses an enumeration of known headers and returns parsed values as QVariants. I would suggest to revert #49966 and have it fixed in QgsNetworkReplyContent to avoid surprises in the future...

**Comment by uclaros** [comment]
@wonder-sk fyi, implemented in #50293
